service/src/test/java/io/camunda/service/AdHocSubProcessActivityServicesTest.java

```diff
@@ -0,0 +1,294 @@
+        .isEqualTo(AD_HOC_SUB_PROCESS_INSTANCE_KEY);
+  }
+
+  @Test
+  public void shouldReturnNewInstanceWithAuthentication() {
+    // given
+    final CamundaAuthentication newAuthentication = CamundaAuthentication.none();
+
+    // when
+    final AdHocSubProcessActivityServices newServices =
+        services.withAuthentication(newAuthentication);
+
+    // then
+    assertThat(newServices).isNotSameAs(services);
+    assertThat(newServices).isInstanceOf(AdHocSubProcessActivityServices.class);
+  }
+
+  @Test
+  public void shouldHandleEmptyElementsList() {
+    // given
+    final AdHocSubProcessActivateActivitiesRequest request =
+        new AdHocSubProcessActivateActivitiesRequest(AD_HOC_SUB_PROCESS_INSTANCE_KEY, List.of(), true);
+
+    final AdHocSubProcessInstructionRecord expectedResponse =
+        new AdHocSubProcessInstructionRecord();
+    final BrokerResponse<AdHocSubProcessInstructionRecord> brokerResponse =
+        new BrokerResponse<>(expectedResponse);
+
+    when(brokerClient.sendRequest(requestCaptor.capture()))
+        .thenReturn(CompletableFuture.completedFuture(brokerResponse));
+
+    // when
+    final AdHocSubProcessInstructionRecord result = services.activateActivities(request).join();
+
+    // then
+    assertThat(result).isEqualTo(expectedResponse);
+
+    final BrokerActivateAdHocSubProcessActivityRequest capturedRequest = requestCaptor.getValue();
+    assertThat(capturedRequest.getRequestWriter().getAdHocSubProcessInstanceKey())
+        .isEqualTo(AD_HOC_SUB_PROCESS_INSTANCE_KEY);
+    assertThat(capturedRequest.getRequestWriter().getActivateElements()).isEmpty();
+  }
+
+  @Test
+  public void shouldPropagateCompletableFutureFromBrokerClient() {
+    // given
+    final AdHocSubProcessActivateActivityReference element =
+        new AdHocSubProcessActivateActivityReference("activity1", Map.of());
+    final AdHocSubProcessActivateActivitiesRequest request =
+        new AdHocSubProcessActivateActivitiesRequest(
+            AD_HOC_SUB_PROCESS_INSTANCE_KEY, List.of(element), true);
+
+    final CompletableFuture<BrokerResponse<AdHocSubProcessInstructionRecord>> brokerFuture =
+        new CompletableFuture<>();
+
+    when(brokerClient.sendRequest(any(BrokerActivateAdHocSubProcessActivityRequest.class)))
+        .thenReturn(brokerFuture);
+
+    // when
+    final CompletableFuture<AdHocSubProcessInstructionRecord> result =
+        services.activateActivities(request);
+
+    // then
+    assertThat(result).isNotCompleted();
+
+    // complete the broker future
+    final AdHocSubProcessInstructionRecord expectedResponse =
+        new AdHocSubProcessInstructionRecord();
+    final BrokerResponse<AdHocSubProcessInstructionRecord> brokerResponse =
+        new BrokerResponse<>(expectedResponse);
+    brokerFuture.complete(brokerResponse);
+
+    assertThat(result.join()).isEqualTo(expectedResponse);
+  }
+
+  @ParameterizedTest
+  @MethodSource("cancelRemainingInstancesProvider")
```

Comment:
🔧 If you change the cancel flag to a primimtive type like suggested in my previous comment you can simplify this and remove the `cancelRemainingInstancesProvider` method:

`@ValueSource(booleans = {true, false})`